elements/lisk-p2p/src/peer_book/base_list.ts

```diff
@@ -13,146 +13,192 @@
+			bucketIdToBucket.set(bucketId, new Map<string, P2PEnhancedPeerInfo>());
 		}
 	}
 
 	public get peerList(): ReadonlyArray<P2PPeerInfo> {
 		const peerListMap: P2PPeerInfo[] = [];
 
-		for (const peerMap of [...this.peerMap.values()]) {
-			for (const peer of [...peerMap.values()]) {
-				peerListMap.push(peer.peerInfo);
+		for (const peerList of [...this.bucketIdToBucket.values()]) {
+			for (const peer of [...peerList.values()]) {
+				// Remove internal fields before sharing
+				peerListMap.push(sanitizeInternalPeerInfo(peer));
 			}
 		}
 
 		return peerListMap;
 	}
 
-	public initPeerInfo = (peerInfo: P2PPeerInfo): CustomPeerInfo => ({
-		peerInfo,
-		dateAdded: new Date(),
-	});
-
-	public getBucket(ipAddress: string): Map<string, CustomPeerInfo> {
+	public calculateBucket(
+		targetAddress: string,
+		sourceAddress?: string,
+	): BucketInfo {
 		const bucketId = getBucketId({
 			secret: this.peerListConfig.secret,
 			peerType: this.peerListConfig.peerType,
-			targetAddress: ipAddress,
-			bucketCount: this.peerListConfig.peerBucketCount,
+			targetAddress,
+			sourceAddress:
+				this.type === PEER_TYPE.NEW_PEER ? sourceAddress : undefined,
+			bucketCount: this.peerListConfig.numOfBuckets,
 		});
 
-		return this.peerMap.get(bucketId) as Map<string, CustomPeerInfo>;
+		return { bucketId, bucket: this.bucketIdToBucket.get(bucketId) as Bucket };
 	}
 
-	public getPeer(peerInfo: P2PPeerInfo): P2PPeerInfo | undefined {
-		const bucket = this.getBucket(peerInfo.ipAddress);
-		const incomingPeerId = peerInfo.peerId;
-		const peer = bucket.get(incomingPeerId);
+	public getPeer(incomingPeerId: string): P2PEnhancedPeerInfo | undefined {
+		const peerInfo = this.peerIdToPeerInfo.get(incomingPeerId);
 
-		return peer ? peer.peerInfo : undefined;
+		if (!peerInfo) {
+			return undefined;
+		}
+
+		return sanitizeInternalPeerInfo(peerInfo);
 	}
 
-	public addPeer(peerInfo: P2PPeerInfo): CustomPeerInfo | undefined {
-		if (this.getPeer(peerInfo)) {
-			throw new ExistingPeerError(peerInfo);
+	public addPeer(
+		incomingPeerInfo: P2PEnhancedPeerInfo,
+	): P2PEnhancedPeerInfo | undefined {
+		const existingPeerInfo = this.getPeer(incomingPeerInfo.peerId);
+		if (existingPeerInfo) {
+			throw new ExistingPeerError(incomingPeerInfo);
 		}
-		const bucket = this.getBucket(peerInfo.ipAddress);
-		const incomingPeerId = peerInfo.peerId;
-		const newPeer = this.initPeerInfo(peerInfo);
-		const evictedPeer = this.makeSpace(peerInfo.ipAddress);
-		bucket.set(incomingPeerId, newPeer);
 
-		// If a peer was evicted in order to make space for the new one, we return its info
+		const { bucketId, bucket } = this.calculateBucket(
+			incomingPeerInfo.ipAddress,
+			this.type === PEER_TYPE.NEW_PEER
+				? incomingPeerInfo.sourceAddress
+				: undefined,
+		);
+
+		// If bucket is full, evict a peer to make space for incoming peer
+		const evictedPeer =
+			bucket.size === this.peerListConfig.bucketSize
```

Comment:
maybe it's better to compare with `bucket.size >= this.peerListConfig.bucketSize` so that even if it overflow, it will resolve?